2021.08.28: Fails to link: error: undefined hidden symbol: rumur::parse(std::__1::basic_istream<char, std::__1::char_traits<char> >&)

```
ld: error: undefined hidden symbol: rumur::parse(std::__1::basic_istream<char, std::__1::char_traits<char> >&)
>>> referenced by main.cc
>>>               murphi2c/CMakeFiles/murphi2c.dir/src/main.cc.o:(main)

ld: error: undefined hidden symbol: rumur::resolve_symbols(rumur::Model&)
>>> referenced by main.cc
>>>               murphi2c/CMakeFiles/murphi2c.dir/src/main.cc.o:(main)

ld: error: undefined hidden symbol: rumur::validate(rumur::Node const&)
>>> referenced by main.cc
>>>               murphi2c/CMakeFiles/murphi2c.dir/src/main.cc.o:(main)

ld: error: undefined hidden symbol: rumur::sanitise_rule_names(rumur::Node&)
>>> referenced by main.cc
>>>               murphi2c/CMakeFiles/murphi2c.dir/src/main.cc.o:(main)
```

clang-11
FreeBSD 13

> The failure occurs during the port build which builds shared libraries.

Ah, I think this was the ingredient I was missing. I can reproduce what I think is a related failure on macOS:

```
$ cmake -DBUILD_SHARED_LIBS=ON .. && cmake --build .
...
d: warning: PIE disabled. Absolute addressing (perhaps -mdynamic-no-pic) not allowed in code signed PIE, but used in __ZN3smt12_GLOBAL__N_110Simplifier9make_trueEv from CMakeFiles/rumur.dir/src/smt/simplify.cc.o. To fix this warning, don't compile with -mdynamic-no-pic or link with -Wl,-no_pie
final section layout:
    __TEXT/__text addr=0x100002340, size=0x0008B750, fileOffset=0x00002340, type=1
    __TEXT/__stubs addr=0x10008DA90, size=0x00000786, fileOffset=0x0008DA90, type=29
    __TEXT/__stub_helper addr=0x10008E218, size=0x000006FA, fileOffset=0x0008E218, type=33
    __TEXT/__const addr=0x10008E920, size=0x00022970, fileOffset=0x0008E920, type=0
    __TEXT/__gcc_except_tab addr=0x1000B1290, size=0x00004218, fileOffset=0x000B1290, type=0
    __TEXT/__cstring addr=0x1000B54A8, size=0x00008F39, fileOffset=0x000B54A8, type=13
    __TEXT/__unwind_info addr=0x1000BE3E4, size=0x00001C10, fileOffset=0x000BE3E4, type=22
    __DATA_CONST/__got addr=0x1000C0000, size=0x00000308, fileOffset=0x000C0000, type=30
    __DATA_CONST/__mod_init_func addr=0x1000C0308, size=0x00000020, fileOffset=0x000C0308, type=34
    __DATA_CONST/__const addr=0x1000C0328, size=0x000035B0, fileOffset=0x000C0328, type=0
    __DATA/__la_symbol_ptr addr=0x1000C4000, size=0x00000A08, fileOffset=0x000C4000, type=28
    __DATA/__data addr=0x1000C4A10, size=0x00000448, fileOffset=0x000C4A10, type=0
    __DATA/__bss addr=0x1000C4E58, size=0x000000A8, fileOffset=0x00000000, type=26
    __DATA/__common addr=0x1000C4F00, size=0x000001C0, fileOffset=0x00000000, type=26
ld: 32-bit RIP relative reference out of range (-4295365849 max is +/-2GB): from __ZN3smt12_GLOBAL__N_110Simplifier9make_trueEv (0x1000614C0) to __ZN5rumur4TrueE@0x000E9AD8 (0x00000000) in '__ZN3smt12_GLOBAL__N_110Simplifier9make_trueEv' from CMakeFiles/rumur.dir/src/smt/simplify.cc.o for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [rumur/rumur] Error 1
make[1]: *** [rumur/CMakeFiles/rumur.dir/all] Error 2
make: *** [all] Error 2
```

This is a little surprising as Rumur's CMake files are pretty vanilla and try not to make any constraining choices wrt static vs shared libraries. I'll try to reproduce the exact failure on FreeBSD next and see if I can root cause this.

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -11,6 +11,12 @@ task:
       freebsd_instance:
         image_family: freebsd-13-0-snap
       install_script: pkg upgrade -y && pkg install -y bash bison cmake gmp libxml2 z3
+    - name: FreeBSD 13.0, shared libraries
+      freebsd_instance:
+        image_family: freebsd-13-0-snap
+      environment:
+        CMAKE_OPTIONS: -DBUILD_SHARED_LIBS=ON
+      install_script: pkg upgrade -y && pkg install -y bash bison cmake gmp libxml2 z3
     - name: FreeBSD 12.2
       freebsd_instance:
         image_family: freebsd-12-2-snap
@@ -146,4 +152,4 @@ task:
   # we limit the test suite to a single thread because the Cirrus CI VMs claim
   # to have 2 CPUs but do not seem to give two concurrent processes enough CPU
   # time and we end up having some of the SMT tests time out
-  test_script: uname -sr && python3 --version && mkdir build && cd build && cmake .. && cmake --build . && cmake --build . -- install && cmake --build . -- check
+  test_script: uname -sr && python3 --version && mkdir build && cd build && cmake ${CMAKE_OPTIONS:-} .. && cmake --build . && cmake --build . -- install && cmake --build . -- check

--- a/librumur/include/rumur/rumur.h
+++ b/librumur/include/rumur/rumur.h
@@ -3,7 +3,7 @@
 #pragma once
 
 #ifndef RUMUR_API
-#define RUMUR_API __attribute__((visibility("hidden")))
+#define RUMUR_API
 #endif
 
 namespace rumur {

